### PR TITLE
feat: add MountFrom and OnMounted to CopyGraphOptions

### LINKED ISSUE #580: Support cross-repository mounting in Copy
We have already have [Mounter](https://pkg.go.dev/oras.land/oras-go/v2@v2.3.0/registry#Mounter) implemented for [Repository](https://pkg.go.dev/oras.land/oras-go/v2@v2.3.0/registry/remote#Repository.Mount), we can consider leveraging it in [Copy](https://pkg.go.dev/oras.land/oras-go/v2@v2.3.0#Copy) and [CopyGraph](https://pkg.go.dev/oras.land/oras-go/v2@v2.3.0#CopyGraph) for better performance.

### PR DESCRIPTION
**Pull request by ktarplee**
Adds MountFrom and OnMounted to CopyGraphOptions.
Allows for trying to mount from multiple repositories.

Closes #580 

I think this is a better approach than my other PR #632

### CODE REVIEW

File: copy.go [hunk 1]
```diff
@@ -105,6 +105,11 @@ type CopyGraphOptions struct {
 	// OnCopySkipped will be called when the sub-DAG rooted by the current node
 	// is skipped.
 	OnCopySkipped func(ctx context.Context, desc ocispec.Descriptor) error
+	// MountFrom returns the candidate repositories that desc may be mounted from.
+	// The OCI references will be tried in turn.  If mounting fails on all of them, then it falls back to a copy.
```
Comment (shizhMSFT) [inline copy.go:109]: nit: formating
```suggestion
	// MountFrom returns the candidate repositories that desc may be mounted from.
	// The OCI references will be tried in turn. If mounting fails on all of them,
	// then it falls back to a copy.
```

File: copy.go [hunk 2]
```diff
@@ -259,12 +264,94 @@ func copyGraph(ctx context.Context, src content.ReadOnlyStorage, dst content.Sto
...
 	}
 
 	return syncutil.Go(ctx, limiter, fn, root)
 }
 
+// mountOrCopyNode enabled cross repository blob mounting.
+// sourceReference is the repository to use for mounting (the mount point).
+// mounter is the destination for the mount (a well-known implementation of this is *registry.Repository representing the target).
+// OnMounted is called (if provided) when the blob is mounted.
+// The original PreCopy hook is called only on copy, and therefore not when the blob is mounted.
```
Comment (shizhMSFT) [inline copy.go:277]: nit: it would be better to wrap the documentation at line width of `80` chars.

File: copy.go [hunk 3]
```diff
@@ -259,12 +264,94 @@ func copyGraph(ctx context.Context, src content.ReadOnlyStorage, dst content.Sto
...
+	sourceRepositories, err := opts.MountFrom(ctx, desc)
+	if err != nil {
+		// Technically this error is not fatal, we can still attempt to copy the node
+		// But for consistency with the other callbacks we bail out.
+		return err
+	}
+
+	if len(sourceRepositories) == 0 {
+		return copyNode(ctx, src, dst, desc, opts)
+	}
```
Comment (shizhMSFT) [inline copy.go:303]: nit: could you refactor this code as it has many paths calling `return copyNode(ctx, src, dst, desc, opts)`?

File: copy.go [hunk 4]
```diff
@@ -259,12 +264,94 @@ func copyGraph(ctx context.Context, src content.ReadOnlyStorage, dst content.Sto
...
+
+	skipContent := errors.New("skip content")
```
Comment (shizhMSFT) [inline copy.go:305]: nit: could you make it a global variable so that it is only created once?

File: copy.go [hunk 5]
```diff
@@ -259,12 +264,94 @@ func copyGraph(ctx context.Context, src content.ReadOnlyStorage, dst content.Sto
...
+	for i, sourceRepository := range sourceRepositories {
+		// try mounting this source repository
+		var mountFailed bool
+		getContent := func() (io.ReadCloser, error) {
+			// the invocation of getContent indicates that mounting has failed
+			mountFailed = true
+
+			if len(sourceRepositories)-1 == i {
+				// this is the last iteration so we need to actually get the content and do the copy
+
```
Comment (shizhMSFT) [inline copy.go:315]: nit: remove extra empty line

**Review by shizhMSFT — APPROVED** [review]
LGTM with suggestions. I think this design is indeed a better approach than #632.

**Comment by shizhMSFT** [comment]
@ktarplee Meanwhile, could you increase the code coverage so that the check can pass?

File: copy.go (hunk 4 above)
Comment (ktarplee) [reply copy.go:307]: I think the benefit of the variable's scope being as small as possible overrides the benefit of the optimization.

File: copy.go (hunk 3 above)
Comment (ktarplee) [reply copy.go]: done

File: copy.go (hunk 2 above)
Comment (ktarplee) [reply copy.go]: done

File: copy.go (hunk 1 above)
Comment (ktarplee) [reply copy.go]: done

**Review by shizhMSFT — APPROVED** [review]
LGTM

File: copy.go [hunk 6]
```diff
@@ -259,12 +265,85 @@ func copyGraph(ctx context.Context, src content.ReadOnlyStorage, dst content.Sto
...
+				if opts.PreCopy != nil {
+					if err := opts.PreCopy(ctx, desc); err != nil {
+						return nil, err
+					}
+				}
+				return src.Fetch(ctx, desc)
+			}
+
+			// We want to return an error that we will test for from mounter.Mount()
+			return nil, skipContent
```
Comment (Wwwsylvia) [inline copy.go:318]: nit: Would something like this be more readable?

```suggestion
			if i < len(sourceRepositories)-1 {
				// If this is not the last one, skip this source and try next one
				// We want to return an error that we will test for from mounter.Mount()
				return nil, skipContent
			}

			// this is the last iteration so we need to actually get the content and do the copy
			// but first call the PreCopy function
			if opts.PreCopy != nil {
				if err := opts.PreCopy(ctx, desc); err != nil {
					return nil, err
				}
			}
			return src.Fetch(ctx, desc)
```

File: copy.go [hunk 7]
```diff
@@ -259,12 +265,85 @@ func copyGraph(ctx context.Context, src content.ReadOnlyStorage, dst content.Sto
...
+		// Technically this error is not fatal, we can still attempt to copy the node
+		// But for consistency with the other callbacks we bail out.
+		return err
+	}
+
+	if len(sourceRepositories) == 0 {
+		return copyNode(ctx, src, dst, desc, opts)
+	}
+
+	skipContent := errors.New("skip content")
```
Comment (Wwwsylvia) [inline copy.go:298]: nit: `skipContent` sounds like this node is skipped, but this error is used to skip the source? Wondering if it should be called something like `skipSource` or `ignoreSource`, etc.?

File: copy.go (hunk 6 above)
Comment (ktarplee) [reply copy.go:318]: I like that better as well.

File: copy.go (hunk 7 above)
Comment (ktarplee) [reply copy.go:298]: Changed to `skipSource`

**Comment by ktarplee** [comment]
@Wwwsylvia @shizhMSFT I pushed my changes.  Sorry about the delay.

**Review by Wwwsylvia — APPROVED** [review]
LGTM

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**